src/ASM/DomainDecomposition.C

```diff
@@ -0,0 +1,787 @@
+      auto dof = sim.getSAM()->getNodeDOFs(i);
+      maxDof += dof.second-dof.first+1;
+      for (auto& it : MLGN)
+        utl::renumber(it, old2new2, false);
+    }
+
+  if (adm.getProcId() < adm.getNoProcs()-1) {
+    adm.send(maxNode, adm.getProcId()+1);
+    adm.send(maxDof, adm.getProcId()+1);
+  }
+
+  for (const auto& it : ghostConnections) {
+    int midx = sim.getLocalPatchIndex(it.master);
+    if (midx < 1)
+      continue;
+
+    std::vector<int> glbNodes;
+    sim.getPatch(midx)->getBoundaryNodes(it.midx, glbNodes);
+    for (size_t i = 0; i < glbNodes.size(); ++i)
+      glbNodes[i] = MLGN[glbNodes[i]-1];
+
+    adm.send(int(glbNodes.size()), getPatchOwner(it.slave));
+    adm.send(glbNodes, getPatchOwner(it.slave));
+  }
+
+#endif
+
+  return true;
+}
+
+
+std::vector<int> DomainDecomposition::setupEquationNumbers(const SIMbase& sim,
+                                                           int pidx, int lidx)
+{
+  std::vector<IntVec> lNodes(sim.getPatch(pidx)->getNoBasis());
+  std::vector<int> result;
+
+  for (size_t block = 0; block < blocks.size(); ++block) {
+    std::set<int> bases;
+    if (block == 0) {
+      for (size_t i = 1; i <= sim.getPatch(pidx)->getNoBasis(); ++i)
+        bases.insert(i);
+    } else
+      bases = utl::getDigits(sim.getSolParams()->getBlock(block-1).basis);
+
+    for (const int& basis : bases) {
+      if (lNodes[basis-1].empty())
+        sim.getPatch(pidx)->getBoundaryNodes(lidx, lNodes[basis-1], basis);
+
+      std::set<int> components;
+      if (block == 0 || sim.getSolParams()->getBlock(block-1).comps == 0) {
+        for (size_t i = 1; i <= sim.getPatch(pidx)->getNoFields(basis); ++i)
+          components.insert(i);
+      } else
+        components = utl::getDigits(sim.getSolParams()->getBlock(block-1).comps);
+
+      for (size_t i = 0; i < lNodes[basis-1].size(); ++i) {
+        int node = lNodes[basis-1][i];
+        for (const int& dof : components) {
+          int eq = sim.getSAM()->getEquation(node, dof);
+          if (eq > 0) {
+            if (block == 0)
+              result.push_back(blocks[block].MLGEQ[eq-1]);
+            else
+              result.push_back(blocks[block].MLGEQ[blocks[block].G2LEQ[eq]-1]);
+          } else
+            result.push_back(0);
+        }
+      }
+    }
+  }
+
+  return result;
+}
+
+
+bool DomainDecomposition::calcGlobalEqNumbers(const ProcessAdm& adm,
+                                              const SIMbase& sim)
+{
+  // defaults in serial and non-parallel runs with MPI enabled builds.
+  blocks[0].minEq = 1;
+  blocks[0].maxEq = sim.getSAM()->getNoEquations();
+  for (size_t i = 1; i < blocks.size(); ++i) {
+    blocks[i].minEq = 1;
+    blocks[i].maxEq = blocks[i].localEqs.size();
+  }
+
+#ifdef HAVE_MPI
+  if (adm.getNoProcs() == 1)
+    return true;
+
+  IFEM::cout << "\tEstablishing global equation numbers" << std::endl;
+
+  auto locLMs = sim.getPatch(1)->getLagrangeMultipliers();
+  std::vector<int> glbLMs;
```

Comment:
OK, vi beholder det som det er nå. Kan evt. gjøre noe med dette (altså de globale LM til SIM) senere en gang, når ting har satt seg hvis man ser noen gevinst (av logiske eller annen natur). Når koden er som kompleks som den nå holder på å bli syns jeg man skal vektlegge litt å gjøre ting logisk riktig også (når det er mulig), fremfor raskeste løsning som i etterkant kan virke forvirrende på andre somgår inn i koden (dette var ment som generell kommentar, ikke myntet på det du har gjort her spesiellt).

Ellers, så får jeg feil når jeg prøvde å kompilere klassen DomainDecomposition. Den liker ikke den innebygde operator= for SlaveOrder klassen (pga. av const referansen der). Men hvis jeg overrider denne slik:

//! \brief Hide ill-formed default assignment operator.
SlaveOrder& operator=(const SlaveOrder&) { return *this; }

så er den mer fornøyd.